lrauv_system_tests/include/lrauv_system_tests/Client.hh

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/*
+ * Development of this module has been funded by the Monterey Bay Aquarium
+ * Research Institute (MBARI) and the David and Lucile Packard Foundation
+ */
+
+#ifndef LRAUV_SYSTEM_TESTS__CLIENT_HH
+#define LRAUV_SYSTEM_TESTS__CLIENT_HH
+
+#include <future>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+
+#include <ignition/common/Console.hh>
+#include <ignition/transport/Node.hh>
+
+namespace lrauv_system_tests
+{
+
+/// Client for Ignition Transport topic-based services.
+/// \tparam RequestMessageT Request message type.
+/// \tparam ResponseMessageT Response message type.
+template<typename RequestMessageT, typename ResponseMessageT>
+class Client
+{
+  /// Constructor.
+  /// \param[in] _node Node to be used for topic advertisement and subscription.
+  /// \param[in] _ns Namespace for the ``requests`` and ``responses`` topics.
+  public: Client(ignition::transport::Node &_node, const std::string& _ns)
+    : requestsCounter(2)
+  {
+    const std::string responseTopicName = _ns + "/responses";
+    if (!_node.Subscribe(responseTopicName, &Client::OnResponse, this))
+    {
+      throw std::runtime_error("Failed to subscribe to " + responseTopicName);
+    }
+    this->requestsPublisher =
+        _node.Advertise<RequestMessageT>(_ns + "/requests");
+  }
+
+  /// Request a response from the remote server.
+  /// \param[in] _request Request message instance.
+  /// \returns a future response message, to be awaited.
+  public: std::future<ResponseMessageT> Request(RequestMessageT _request)
```

Comment:
Not `const`, I have to mutate it right below on line 64. It _could_ be a mutable reference though. I have a slight preference towards not mutating the caller's scope (it's less surprising that way), but I can be convinced otherwise.

Reply:
That's fine. In any case this is stuff that we can change dynamically.